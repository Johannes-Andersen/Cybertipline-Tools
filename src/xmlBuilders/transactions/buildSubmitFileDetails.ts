import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import type { SubmitFileDetails } from '../../types';
import { buildDeviceId } from '../common/buildDeviceId';
import { buildIpCapture } from '../common/buildIpCapture';
import { buildFileDetails } from '../fileDetails/buildFileDetails';
import { buildOriginalFileHash } from '../fileDetails/buildOriginalFileHash';

const parser = new XMLParser({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  allowBooleanAttributes: true,
});

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
  suppressEmptyNode: true,
  processEntities: false,
});

export const buildSubmitFileDetails = (
  fileDetails: SubmitFileDetails,
  keyName = 'fileDetails',
): string => {
  const { originalFileHash, ipCaptureEvent, deviceId, details } = fileDetails;

  const originalFileHashes = originalFileHash?.map(
    (e) => parser.parse(buildOriginalFileHash(e)).originalFileHash,
  );
  const deviceIds = deviceId?.map(
    (e) => parser.parse(buildDeviceId(e)).deviceId,
  );
  const ipCaptureEvents = ipCaptureEvent
    ? parser.parse(buildIpCapture(ipCaptureEvent)).ipCaptureEvent
    : undefined;
  const detail = details?.map((e) => parser.parse(buildFileDetails(e)).details);

  return builder.build({
    '?xml': {
      '@_version': '1.0',
      '@_encoding': 'UTF-8',
    },
    [keyName]: {
      '@_xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
      '@_xsi:noNamespaceSchemaLocation':
        'https://report.cybertip.org/ispws/xsd',
      reportId: fileDetails.reportId,
      fileId: fileDetails.fileId,
      originalFileName: fileDetails.originalFileName,
      locationOfFile: fileDetails.locationOfFile,
      fileViewedByEsp: fileDetails.fileViewedByEsp,
      exifViewedByEsp: fileDetails.exifViewedByEsp,
      publiclyAvailable: fileDetails.publiclyAvailable,
      fileRelevance: fileDetails.fileRelevance,
      fileAnnotations: Object.fromEntries(
        Object.entries(fileDetails.fileAnnotations ?? {})
          .filter(([_, value]) => value)
          .map(([key]) => [key, '']),
      ),
      industryClassification: fileDetails.industryClassification,
      originalFileHash: originalFileHashes,
      ipCaptureEvent: ipCaptureEvents,
      deviceId: deviceIds,
      details: detail,
      additionalInfo: fileDetails.additionalInfo,
    },
  });
};
