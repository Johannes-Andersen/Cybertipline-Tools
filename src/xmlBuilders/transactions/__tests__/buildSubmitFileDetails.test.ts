import { describe, expect, it } from 'vitest';
import {
  DeviceEventName,
  DeviceIdType,
  FileAnnotations,
  FileDetailType,
  FileHashType,
  FileRelevance,
  IPEventName,
  IndustryClassification,
} from '../../../types';
import type { SubmitFileDetails } from '../../../types';
import { buildSubmitFileDetails } from '../buildSubmitFileDetails';

describe('buildSubmitFileDetails', () => {
  it('should build file details with all fields', () => {
    const fileDetails: SubmitFileDetails = {
      reportId: 123456,
      fileId: 'FILE789',
      fileName: 'stored_evidence.jpg',
      originalFileName: 'original_evidence.jpg',
      locationOfFile: 'https://platform.com/files/evidence.jpg',
      fileViewedByEsp: true,
      exifViewedByEsp: true,
      publiclyAvailable: false,
      fileRelevance: FileRelevance.Reported,
      fileAnnotations: {
        [FileAnnotations.Viral]: true,
        [FileAnnotations.LiveStreaming]: true,
        [FileAnnotations.GenerativeAI]: false,
      },
      industryClassification: IndustryClassification.A1,
      originalFileHash: [
        {
          value:
            '5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8',
          hashType: FileHashType.SHA1,
        },
        {
          value: 'e86fdc2283aff4179685f2a918a769eb',
          hashType: FileHashType.MD5,
        },
      ],
      ipCaptureEvent: {
        ipAddress: '192.168.5.100',
        eventName: IPEventName.Upload,
        dateTime: new Date('2024-01-15T16:30:00Z'),
        possibleProxy: false,
        port: 443,
      },
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '354123456789012',
          eventName: DeviceEventName.Upload,
          dateTime: new Date('2024-01-15T16:30:00Z'),
        },
      ],
      details: [
        {
          type: FileDetailType.EXIF,
          valuePair: [
            {
              name: 'Make',
              value: 'Canon',
            },
            {
              name: 'Model',
              value: 'EOS R5',
            },
          ],
        },
        {
          type: FileDetailType.HASH,
          valuePair: [
            {
              name: 'PhotoDNA',
              value: 'hash_value_1',
            },
          ],
        },
      ],
      additionalInfo: [
        'File was uploaded via mobile app',
        'User attempted to delete file after upload',
      ],
    };

    const result = buildSubmitFileDetails(fileDetails);

    expect(result).toMatchSnapshot();
  });

  it('should build file details with only required fields', () => {
    const fileDetails: SubmitFileDetails = {
      reportId: 123456,
      fileId: 'FILE789',
    };

    const result = buildSubmitFileDetails(fileDetails);

    expect(result).toMatchSnapshot();
  });

  it('should build file details with file annotations and relevance', () => {
    const fileDetails: SubmitFileDetails = {
      reportId: 123456,
      fileId: 'FILE789',
      fileRelevance: FileRelevance.SupplementalReported,
      fileAnnotations: {
        [FileAnnotations.PotentialMeme]: true,
        [FileAnnotations.Viral]: true,
      },
      additionalInfo: ['File appears to be a viral meme'],
    };

    const result = buildSubmitFileDetails(fileDetails);

    expect(result).toMatchSnapshot();
  });

  it('should build file details with technical metadata', () => {
    const fileDetails: SubmitFileDetails = {
      reportId: 123456,
      fileId: 'FILE789',
      fileViewedByEsp: true,
      exifViewedByEsp: true,
      details: [
        {
          type: FileDetailType.EXIF,
          valuePair: [
            {
              name: 'DateTime',
              value: '2024:01:15 16:30:00',
            },
            {
              name: 'GPSLatitude',
              value: '40.7128° N',
            },
            {
              name: 'GPSLongitude',
              value: '74.0060° W',
            },
          ],
        },
      ],
      deviceId: [
        {
          idType: DeviceIdType.ICCID,
          idValue: '89012345678901234567',
          eventName: DeviceEventName.Upload,
        },
      ],
    };

    const result = buildSubmitFileDetails(fileDetails);

    expect(result).toMatchSnapshot();
  });
});
