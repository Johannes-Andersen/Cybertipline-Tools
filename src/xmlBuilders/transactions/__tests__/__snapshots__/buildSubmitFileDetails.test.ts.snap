// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildSubmitFileDetails > should build file details with all fields 1`] = `"<?xml version="1.0" encoding="UTF-8"?><fileDetails xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><reportId>123456</reportId><fileId>FILE789</fileId><originalFileName>original_evidence.jpg</originalFileName><locationOfFile>https://platform.com/files/evidence.jpg</locationOfFile><fileViewedByEsp>true</fileViewedByEsp><exifViewedByEsp>true</exifViewedByEsp><publiclyAvailable>false</publiclyAvailable><fileRelevance>Reported</fileRelevance><fileAnnotations><viral/><liveStreaming/></fileAnnotations><industryClassification>A1</industryClassification><originalFileHash hashType="SHA1">5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8</originalFileHash><originalFileHash hashType="MD5">e86fdc2283aff4179685f2a918a769eb</originalFileHash><ipCaptureEvent><ipAddress>192.168.5.100</ipAddress><eventName>Upload</eventName><dateTime>2024-01-15T16:30:00.000Z</dateTime><possibleProxy>false</possibleProxy><port>443</port></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>354123456789012</idValue><eventName>Upload</eventName><dateTime>2024-01-15T16:30:00.000Z</dateTime></deviceId><details type="EXIF"><nameValuePair><name>Make</name><value>Canon</value></nameValuePair><nameValuePair><name>Model</name><value>EOS R5</value></nameValuePair></details><details type="HASH"><nameValuePair><name>PhotoDNA</name><value>hash_value_1</value></nameValuePair></details><additionalInfo>File was uploaded via mobile app</additionalInfo><additionalInfo>User attempted to delete file after upload</additionalInfo></fileDetails>"`;

exports[`buildSubmitFileDetails > should build file details with file annotations and relevance 1`] = `"<?xml version="1.0" encoding="UTF-8"?><fileDetails xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><reportId>123456</reportId><fileId>FILE789</fileId><fileRelevance>Supplemental Reported</fileRelevance><fileAnnotations><potentialMeme/><viral/></fileAnnotations><additionalInfo>File appears to be a viral meme</additionalInfo></fileDetails>"`;

exports[`buildSubmitFileDetails > should build file details with only required fields 1`] = `"<?xml version="1.0" encoding="UTF-8"?><fileDetails xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><reportId>123456</reportId><fileId>FILE789</fileId><fileAnnotations/></fileDetails>"`;

exports[`buildSubmitFileDetails > should build file details with technical metadata 1`] = `"<?xml version="1.0" encoding="UTF-8"?><fileDetails xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><reportId>123456</reportId><fileId>FILE789</fileId><fileViewedByEsp>true</fileViewedByEsp><exifViewedByEsp>true</exifViewedByEsp><fileAnnotations/><deviceId><idType>ICCID</idType><idValue>89012345678901234567</idValue><eventName>Upload</eventName></deviceId><details type="EXIF"><nameValuePair><name>DateTime</name><value>2024:01:15 16:30:00</value></nameValuePair><nameValuePair><name>GPSLatitude</name><value>40.7128° N</value></nameValuePair><nameValuePair><name>GPSLongitude</name><value>74.0060° W</value></nameValuePair></details></fileDetails>"`;
